From 70daff75a83069724be9274a614caa5a6e308aef Mon Sep 17 00:00:00 2001
From: brandnkdavis <brandnkdavis@gmail.com>
Date: Sun, 10 May 2026 21:33:10 -0500
Subject: [PATCH 09/17] sepolicy: qti: Allow rules for mdmfeature

Change-Id: Ieef823f2ecea4dfb055b6fd6a6bb83bf0a889a3d
(cherry picked from commit b070c0c42c410ad0e19da22a76d1ef8cf094ad95)

Ledger: hardware_oplus.md Accepted Group A row 3 (b070c0c4)
Change-Id: Id76ae4133fce864e55285b167ca05ec47a8e8e77
(cherry picked from commit 02f9bf6ac46e7caae27fa050a04197412bc37200)
---
 sepolicy/qti/vendor/device.te            |  3 +++
 sepolicy/qti/vendor/file_contexts        |  4 ++++
 sepolicy/qti/vendor/oplus_mdm_feature.te | 23 +++++++++++++++++++++++
 sepolicy/qti/vendor/property_contexts    | 10 ++++++----
 4 files changed, 36 insertions(+), 4 deletions(-)
 create mode 100644 sepolicy/qti/vendor/oplus_mdm_feature.te

diff --git a/sepolicy/qti/vendor/device.te b/sepolicy/qti/vendor/device.te
index 25040219..d758893b 100644
--- a/sepolicy/qti/vendor/device.te
+++ b/sepolicy/qti/vendor/device.te
@@ -19,6 +19,9 @@ type fingerprint_device, dev_type;
 # IR
 type oplus_consumer_ir_device, dev_type;
 
+# MDM feature
+type oplus_mdmfeature_device, dev_type;
+
 # NVRAM
 type nvram_device, dev_type;
 
diff --git a/sepolicy/qti/vendor/file_contexts b/sepolicy/qti/vendor/file_contexts
index ed7ba358..3351c500 100644
--- a/sepolicy/qti/vendor/file_contexts
+++ b/sepolicy/qti/vendor/file_contexts
@@ -94,6 +94,10 @@
 /dev/block/platform/soc/1d84000\.ufshc/by-name/mdmddr             u:object_r:vendor_efs_boot_dev:s0
 /dev/block/platform/soc/4804000\.ufshc/by-name/mdm1oemnvbktmp     u:object_r:vendor_modem_efs_partition_device:s0
 
+# MDM feature
+/(odm|vendor/odm)/bin/hw/mdm_feature    u:object_r:oplus_mdm_feature_exec:s0
+/dev/mdmfeature                         u:object_r:oplus_mdmfeature_device:s0
+
 # NFC
 /dev/pn553    u:object_r:nfc_device:s0
 
diff --git a/sepolicy/qti/vendor/oplus_mdm_feature.te b/sepolicy/qti/vendor/oplus_mdm_feature.te
new file mode 100644
index 00000000..e2b7313d
--- /dev/null
+++ b/sepolicy/qti/vendor/oplus_mdm_feature.te
@@ -0,0 +1,23 @@
+type oplus_mdm_feature, domain;
+binder_use(oplus_mdm_feature)
+
+type oplus_mdm_feature_exec, exec_type, vendor_file_type, file_type;
+init_daemon_domain(oplus_mdm_feature)
+
+allow oplus_mdm_feature block_device:dir r_dir_perms;
+allow oplus_mdm_feature vendor_modem_efs_partition_device:blk_file r_file_perms;
+allow oplus_mdm_feature vendor_reserve_partition:blk_file rw_file_perms;
+
+allow oplus_mdm_feature oplus_mdmfeature_device:chr_file rw_file_perms;
+
+allow oplus_mdm_feature vendor_sysfs_soc_sensitive:file r_file_perms;
+
+allow oplus_mdm_feature mnt_vendor_file:dir search;
+allow oplus_mdm_feature oplus_reserve_file_type:dir search;
+
+r_dir_file(oplus_mdm_feature, vendor_proc_engineer)
+rw_dir_file(oplus_mdm_feature, oplus_reserve_radio_file_type)
+rw_dir_file(oplus_mdm_feature, vendor_olog_file)
+
+set_prop(oplus_mdm_feature, vendor_nw_exported_system_prop)
+set_prop(oplus_mdm_feature, vendor_radio_prop)
diff --git a/sepolicy/qti/vendor/property_contexts b/sepolicy/qti/vendor/property_contexts
index 725e170f..64b72798 100644
--- a/sepolicy/qti/vendor/property_contexts
+++ b/sepolicy/qti/vendor/property_contexts
@@ -36,10 +36,12 @@ vendor.modem_dump_action    u:object_r:vendor_modem_prop:s0
 vendor.oplus.nfc.fw.version    u:object_r:vendor_nfc_nq_prop:s0
 
 # Radio
-odm.ril.radio.status.          u:object_r:vendor_radio_prop:s0
-persist.vendor.oplus.radio.    u:object_r:vendor_radio_prop:s0
-ro.vendor.oplus.radio.         u:object_r:vendor_radio_prop:s0
-vendor.oplus.radio.            u:object_r:vendor_radio_prop:s0
+odm.ril.radio.status.                 u:object_r:vendor_radio_prop:s0
+persist.vendor.oplus.radio.           u:object_r:vendor_radio_prop:s0
+ro.vendor.oplus.mdmfeature            u:object_r:vendor_radio_prop:s0
+ro.vendor.oplus.radio.                u:object_r:vendor_radio_prop:s0
+ro.vendor.oplus.uniform_single_sim    u:object_r:vendor_radio_prop:s0
+vendor.oplus.radio.                   u:object_r:vendor_radio_prop:s0
 
 # RIL
 vendor.gsm.phoneserial    u:object_r:vendor_nw_exported_system_prop:s0
